// astro/dist/build/page.js
import _path from "path";
import {fileURLToPath} from "url";
import {convertMatchToLocation, validateGetStaticPathsModule, validateGetStaticPathsResult} from "../util.js";
import {generatePaginateFunction} from "./paginate.js";
import {generateRssFunction} from "./rss.js";
async function getStaticPathsForPage({
  astroConfig,
  astroRuntime,
  snowpackRuntime,
  route,
  logging
}) {
  const location = convertMatchToLocation(route, astroConfig);
  const mod = await snowpackRuntime.importModule(location.snowpackURL);
  validateGetStaticPathsModule(mod);
  const [rssFunction, rssResult] = generateRssFunction(astroConfig.buildOptions.site, route);
  const staticPaths = await astroRuntime.getStaticPaths(route.component, mod, {
    paginate: generatePaginateFunction(route),
    rss: rssFunction
  });
  validateGetStaticPathsResult(staticPaths, logging);
  return {
    paths: staticPaths.map((staticPath) => staticPath.params && route.generate(staticPath.params)).filter(Boolean),
    rss: rssResult
  };
}
function formatOutFile(path, pageUrlFormat) {
  if (path === "/404") {
    return "/404.html";
  }
  if (path === "/") {
    return "/index.html";
  }
  if (pageUrlFormat === "directory") {
    return _path.posix.join(path, "/index.html");
  }
  return `${path}.html`;
}
async function buildStaticPage({astroConfig, buildState, path, route, astroRuntime}) {
  const location = convertMatchToLocation(route, astroConfig);
  const normalizedPath = astroConfig.devOptions.trailingSlash === "never" ? path : path.endsWith("/") ? path : `${path}/`;
  const result = await astroRuntime.load(normalizedPath);
  if (result.statusCode !== 200) {
    let err = result.error;
    if (!(err instanceof Error))
      err = new Error(err);
    err.filename = fileURLToPath(location.fileURL);
    throw err;
  }
  buildState[formatOutFile(path, astroConfig.buildOptions.pageUrlFormat)] = {
    srcPath: location.fileURL,
    contents: result.contents,
    contentType: "text/html",
    encoding: "utf8"
  };
}
export {
  buildStaticPage,
  getStaticPathsForPage
};
