// @astrojs/parser/dist/parse/read/expression.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};
__markAsModule(exports);
__export(exports, {
  default: () => read_expression,
  parse_expression_at: () => parse_expression_at
});
var import__2 = __toModule(require("../index.js"));
function peek_char(state) {
  return state.source[state.index];
}
function peek_nonwhitespace(state) {
  let index = state.index;
  do {
    let char = state.source[index];
    if (!/\s/.test(char)) {
      return char;
    }
    index++;
  } while (index < state.source.length);
}
function next_char(state) {
  return state.source[state.index++];
}
function in_bounds(state) {
  return state.index < state.source.length;
}
function consume_string(state, stringChar) {
  let inEscape;
  do {
    const char = next_char(state);
    if (inEscape) {
      inEscape = false;
    } else if (char === "\\") {
      inEscape = true;
    } else if (char === stringChar) {
      break;
    }
  } while (in_bounds(state));
}
function consume_multiline_comment(state) {
  do {
    const char = next_char(state);
    if (char === "*" && peek_char(state) === "/") {
      break;
    }
  } while (in_bounds(state));
}
function consume_line_comment(state) {
  do {
    const char = next_char(state);
    if (char === "\n") {
      break;
    }
  } while (in_bounds(state));
}
const voidElements = new Set(["area", "base", "br", "col", "command", "embed", "hr", "img", "input", "keygen", "link", "meta", "param", "source", "track", "wbr"]);
function consume_tag(state) {
  const start = state.index - 1;
  let tagName = "";
  let inTag = false;
  let inStart = true;
  let selfClosed = false;
  let inClose = false;
  let bracketIndex = 1;
  do {
    const char = next_char(state);
    switch (char) {
      case "'":
      case '"':
      case "`": {
        consume_string(state, char);
        break;
      }
      case "<": {
        inTag = false;
        tagName = "";
        if (peek_nonwhitespace(state) === "/") {
          inClose = true;
          bracketIndex--;
        } else {
          inStart = true;
          bracketIndex++;
        }
        break;
      }
      case ">": {
        if (!inStart && !inClose) {
          break;
        }
        bracketIndex--;
        const addExpectedBrackets = !voidElements.has(tagName.toLowerCase()) && !selfClosed && !inClose;
        if (addExpectedBrackets) {
          bracketIndex += 2;
        }
        inTag = false;
        selfClosed = false;
        inStart = false;
        inClose = false;
        break;
      }
      case " ": {
        inTag = true;
        break;
      }
      case "/": {
        if (inStart) {
          selfClosed = true;
        }
        break;
      }
      default: {
        if (!inTag) {
          tagName += char;
        }
        break;
      }
    }
    if (state.curlyCount <= 0) {
      break;
    }
    if (bracketIndex === 0) {
      break;
    }
  } while (in_bounds(state));
  const source = state.source.substring(start, state.index);
  const ast = (0, import__2.default)(source);
  state.parser.feature_flags |= ast.meta.features;
  const fragment = ast.html;
  return fragment;
}
function consume_expression(parser, source, start) {
  const expr = {
    type: "Expression",
    start,
    end: Number.NaN,
    codeChunks: [],
    children: []
  };
  let codeStart = start;
  const state = {
    source,
    start,
    index: start,
    curlyCount: 1,
    bracketCount: 0,
    root: expr,
    parser
  };
  do {
    const char = next_char(state);
    switch (char) {
      case "{": {
        state.curlyCount++;
        break;
      }
      case "}": {
        state.curlyCount--;
        break;
      }
      case "<": {
        const chunk = source.substring(codeStart, state.index - 1);
        expr.codeChunks.push(chunk);
        const tag = consume_tag(state);
        expr.children.push(tag);
        codeStart = state.index;
        break;
      }
      case "'":
      case '"':
      case "`": {
        consume_string(state, char);
        break;
      }
      case "/": {
        switch (peek_char(state)) {
          case "/": {
            consume_line_comment(state);
            break;
          }
          case "*": {
            consume_multiline_comment(state);
            break;
          }
        }
      }
    }
  } while (in_bounds(state) && state.curlyCount > 0);
  expr.end = state.index - 1;
  if (expr.children.length || !expr.codeChunks.length) {
    expr.codeChunks.push(source.substring(codeStart, expr.end));
  }
  return expr;
}
const parse_expression_at = (parser, source, index) => {
  const expression = consume_expression(parser, source, index);
  return expression;
};
function read_expression(parser) {
  try {
    const expression = parse_expression_at(parser, parser.template, parser.index);
    parser.index = expression.end;
    return expression;
  } catch (err) {
    parser.acorn_error(err);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  parse_expression_at
});
