// astro/dist/compiler/transform/head.js
import {EndOfHead} from "./util/end-of-head.js";
function head_default(opts) {
  let hasComponents = false;
  let isHmrEnabled = typeof opts.compileOptions.hmrPort !== "undefined" && opts.compileOptions.mode === "development";
  const eoh = new EndOfHead();
  return {
    visitors: {
      html: {
        Fragment: {
          enter(node) {
            eoh.enter(node);
          },
          leave(node) {
            eoh.leave(node);
          }
        },
        InlineComponent: {
          enter(node) {
            if (hasComponents) {
              return;
            }
            eoh.enter(node);
            if (node.attributes && node.attributes.some(({name}) => name == null ? void 0 : name.startsWith("client:"))) {
              hasComponents = true;
              return;
            }
            const [_name, kind] = node.name.split(":");
            if (kind) {
              hasComponents = true;
            }
          },
          leave(node) {
            eoh.leave(node);
          }
        },
        Element: {
          enter(node) {
            eoh.enter(node);
          },
          leave(node) {
            eoh.leave(node);
          }
        }
      }
    },
    async finalize() {
      const children = [];
      children.push({
        start: 0,
        end: 0,
        type: "Fragment",
        children: [
          {
            start: 0,
            end: 0,
            type: "Expression",
            codeChunks: ["Astro.pageCSS.map(css => (", "))"],
            children: [
              {
                type: "Element",
                name: "link",
                attributes: [
                  {
                    name: "rel",
                    type: "Attribute",
                    value: [
                      {
                        type: "Text",
                        raw: "stylesheet",
                        data: "stylesheet"
                      }
                    ]
                  },
                  {
                    name: "href",
                    type: "Attribute",
                    value: [
                      {
                        start: 0,
                        end: 0,
                        type: "MustacheTag",
                        expression: {
                          start: 0,
                          end: 0,
                          type: "Expression",
                          codeChunks: ["css"],
                          children: []
                        }
                      }
                    ]
                  }
                ],
                start: 0,
                end: 0,
                children: []
              }
            ]
          },
          {
            start: 0,
            end: 0,
            type: "Expression",
            codeChunks: ["Astro.pageScripts.map(script => (", "))"],
            children: [
              {
                start: 0,
                end: 0,
                type: "Expression",
                codeChunks: ["script.src ? (", ") : (", ")"],
                children: [
                  {
                    type: "Element",
                    name: "script",
                    attributes: [
                      {
                        type: "Attribute",
                        name: "type",
                        value: [
                          {
                            type: "Text",
                            raw: "module",
                            data: "module"
                          }
                        ]
                      },
                      {
                        type: "Attribute",
                        name: "src",
                        value: [
                          {
                            start: 0,
                            end: 0,
                            type: "MustacheTag",
                            expression: {
                              start: 0,
                              end: 0,
                              type: "Expression",
                              codeChunks: ["script.src"],
                              children: []
                            }
                          }
                        ]
                      },
                      {
                        type: "Attribute",
                        name: "data-astro",
                        value: [
                          {
                            type: "Text",
                            raw: "hoist",
                            data: "hoist"
                          }
                        ]
                      }
                    ],
                    start: 0,
                    end: 0,
                    children: []
                  },
                  {
                    type: "Element",
                    name: "script",
                    attributes: [
                      {
                        type: "Attribute",
                        name: "type",
                        value: [
                          {
                            type: "Text",
                            raw: "module",
                            data: "module"
                          }
                        ]
                      },
                      {
                        type: "Attribute",
                        name: "data-astro",
                        value: [
                          {
                            type: "Text",
                            raw: "hoist",
                            data: "hoist"
                          }
                        ]
                      }
                    ],
                    start: 0,
                    end: 0,
                    children: [
                      {
                        start: 0,
                        end: 0,
                        type: "MustacheTag",
                        expression: {
                          start: 0,
                          end: 0,
                          type: "Expression",
                          codeChunks: ["script.content"],
                          children: []
                        }
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      });
      if (hasComponents) {
        children.push({
          type: "Element",
          name: "style",
          attributes: [{name: "type", type: "Attribute", value: [{type: "Text", raw: "text/css", data: "text/css"}]}],
          start: 0,
          end: 0,
          children: [
            {
              start: 0,
              end: 0,
              type: "Text",
              data: "astro-root, astro-fragment { display: contents; }",
              raw: "astro-root, astro-fragment { display: contents; }"
            }
          ]
        });
      }
      if (isHmrEnabled) {
        const {hmrPort} = opts.compileOptions;
        children.push({
          type: "Element",
          name: "script",
          attributes: [],
          children: [{type: "Text", data: `window.HMR_WEBSOCKET_PORT = ${hmrPort};`, start: 0, end: 0}],
          start: 0,
          end: 0
        }, {
          type: "Element",
          name: "script",
          attributes: [
            {type: "Attribute", name: "type", value: [{type: "Text", data: "module", start: 0, end: 0}], start: 0, end: 0},
            {type: "Attribute", name: "src", value: [{type: "Text", data: "/_snowpack/hmr-client.js", start: 0, end: 0}], start: 0, end: 0}
          ],
          children: [],
          start: 0,
          end: 0
        });
      }
      if (eoh.foundHeadOrHtmlElement || eoh.foundHeadAndBodyContent) {
        const topLevelFragment = {
          start: 0,
          end: 0,
          type: "Fragment",
          children
        };
        eoh.append(topLevelFragment);
      }
    }
  };
}
export {
  head_default as default
};
