// astro/dist/compiler/transform/index.js
import {walk} from "estree-walker";
import transformStyles from "./styles.js";
import transformDoctype from "./doctype.js";
import transformModuleScripts from "./module-scripts.js";
import transformCodeBlocks from "./prism.js";
import transformHead from "./head.js";
function addVisitor(visitor, collection, nodeName, event) {
  if (typeof visitor[event] !== "function")
    return;
  if (!collection[event])
    collection[event] = new Map();
  const visitors = collection[event].get(nodeName) || [];
  visitors.push(visitor[event]);
  collection[event].set(nodeName, visitors);
}
function collectVisitors(transformer, htmlVisitors, cssVisitors, finalizers) {
  if (transformer.visitors) {
    if (transformer.visitors.html) {
      for (const [nodeName, visitor] of Object.entries(transformer.visitors.html)) {
        addVisitor(visitor, htmlVisitors, nodeName, "enter");
        addVisitor(visitor, htmlVisitors, nodeName, "leave");
      }
    }
    if (transformer.visitors.css) {
      for (const [nodeName, visitor] of Object.entries(transformer.visitors.css)) {
        addVisitor(visitor, cssVisitors, nodeName, "enter");
        addVisitor(visitor, cssVisitors, nodeName, "leave");
      }
    }
  }
  finalizers.push(transformer.finalize);
}
function createVisitorCollection() {
  return {
    enter: new Map(),
    leave: new Map()
  };
}
function walkAstWithVisitors(tmpl, collection) {
  walk(tmpl, {
    enter(node, parent, key, index) {
      if (collection.enter.has(node.type)) {
        const fns = collection.enter.get(node.type) || [];
        for (let fn of fns) {
          fn.call(this, node, parent, key, index);
        }
      }
    },
    leave(node, parent, key, index) {
      if (collection.leave.has(node.type)) {
        const fns = collection.leave.get(node.type) || [];
        for (let fn of fns) {
          fn.call(this, node, parent, key, index);
        }
      }
    }
  });
}
async function transform(ast, opts) {
  const htmlVisitors = createVisitorCollection();
  const cssVisitors = createVisitorCollection();
  const finalizers = [];
  const optimizers = [transformHead(opts), transformStyles(opts), transformDoctype(opts), transformModuleScripts(opts), transformCodeBlocks(ast.module)];
  for (const optimizer of optimizers) {
    collectVisitors(optimizer, htmlVisitors, cssVisitors, finalizers);
  }
  (ast.css || []).map((css) => walkAstWithVisitors(css, cssVisitors));
  walkAstWithVisitors(ast.html, htmlVisitors);
  await Promise.all(finalizers.map((fn) => fn()));
}
export {
  transform
};
