// astro/dist/compiler/transform/module-scripts.js
import {getAttrValue, setAttrValue} from "../../ast.js";
function module_scripts_default({compileOptions, filename}) {
  const {astroConfig} = compileOptions;
  const fileUrl = new URL(`file://${filename}`);
  return {
    visitors: {
      html: {
        Element: {
          enter(node) {
            let name = node.name;
            if (name !== "script") {
              return;
            }
            let type = getAttrValue(node.attributes, "type");
            if (type !== "module") {
              return;
            }
            let src = getAttrValue(node.attributes, "src");
            if (!src) {
              return;
            }
            if (src.startsWith("/")) {
              return;
            }
            try {
              new URL(src);
              return;
            } catch (err) {
            }
            const srcUrl = new URL(src, fileUrl);
            const absoluteUrl = `/_astro/${srcUrl.href.replace(astroConfig.projectRoot.href, "")}`;
            setAttrValue(node.attributes, "src", absoluteUrl);
          }
        }
      }
    },
    async finalize() {
    }
  };
}
export {
  module_scripts_default as default
};
