// astro/dist/compiler/transform/styles.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {enumerable: true, configurable: true, writable: true, value}) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
import crypto from "crypto";
import {createRequire} from "module";
import path from "path";
import autoprefixer from "autoprefixer";
import postcss from "postcss";
import postcssKeyframes from "postcss-icss-keyframes";
import findUp from "find-up";
import sass from "sass";
import {error} from "../../logger.js";
import astroScopedStyles, {NEVER_SCOPED_TAGS} from "./postcss-scoped-styles/index.js";
import slash from "slash";
const getStyleType = new Map([
  [".css", "css"],
  [".pcss", "postcss"],
  [".sass", "sass"],
  [".scss", "scss"],
  ["css", "css"],
  ["sass", "sass"],
  ["scss", "scss"],
  ["text/css", "css"],
  ["text/sass", "sass"],
  ["text/scss", "scss"]
]);
function hashFromFilename(filename) {
  const hash = crypto.createHash("sha256");
  return hash.update(slash(filename)).digest("base64").toString().replace(/[^A-Za-z0-9-]/g, "").substr(0, 8);
}
const miniCache = {
  nodeModules: new Map()
};
function hasClass(classList, className) {
  if (!className)
    return false;
  for (const c of classList.split(" ")) {
    if (className === c.trim())
      return true;
  }
  return false;
}
async function transformStyle(code, {logging, type, filename, scopedClass, tailwindConfig, global}) {
  let styleType = "css";
  if (type) {
    styleType = getStyleType.get(type) || styleType;
  }
  let includePaths = [path.dirname(filename)];
  const cachedNodeModulesDir = miniCache.nodeModules.get(filename);
  if (cachedNodeModulesDir) {
    includePaths.push(cachedNodeModulesDir);
  } else {
    const nodeModulesDir = await findUp("node_modules", {type: "directory", cwd: path.dirname(filename)});
    if (nodeModulesDir) {
      miniCache.nodeModules.set(filename, nodeModulesDir);
      includePaths.push(nodeModulesDir);
    }
  }
  let css = "";
  switch (styleType) {
    case "css": {
      css = code;
      break;
    }
    case "sass":
    case "scss": {
      css = sass.renderSync({data: code, includePaths, indentedSyntax: styleType === "sass"}).css.toString("utf8");
      break;
    }
    default: {
      throw new Error(`Unsupported: <style lang="${styleType}">`);
    }
  }
  const postcssPlugins = [];
  if (tailwindConfig) {
    try {
      const require2 = createRequire(import.meta.url);
      const tw = require2.resolve("tailwindcss", {paths: [import.meta.url, process.cwd()]});
      postcssPlugins.push(require2(tw)(tailwindConfig));
    } catch (err) {
      error(logging, "transform", err);
      throw new Error(`tailwindcss not installed. Try running \`npm install tailwindcss\` and trying again.`);
    }
  }
  if (!global) {
    postcssPlugins.push(astroScopedStyles({className: scopedClass}));
    postcssPlugins.push(postcssKeyframes({
      generateScopedName(keyframesName) {
        return `${keyframesName}-${scopedClass}`;
      }
    }));
  }
  postcssPlugins.push(autoprefixer());
  css = await postcss(postcssPlugins).process(css, {from: filename, to: void 0}).then((result) => result.css);
  return {css, type: styleType};
}
function injectScopedClassAttribute(node, scopedClass, attribute = "class") {
  if (!node.attributes)
    node.attributes = [];
  const classIndex = node.attributes.findIndex(({name}) => name === attribute);
  if (classIndex === -1) {
    node.attributes.push({start: -1, end: -1, type: "Attribute", name: attribute, value: [{type: "Text", raw: scopedClass, data: scopedClass}]});
  } else {
    const attr = node.attributes[classIndex];
    for (let k = 0; k < attr.value.length; k++) {
      if (attr.value[k].type === "Text") {
        if (!hasClass(attr.value[k].data, scopedClass)) {
          attr.value[k].raw += " " + scopedClass;
          attr.value[k].data += " " + scopedClass;
        }
      } else if (attr.value[k].type === "MustacheTag" && attr.value[k]) {
        if (!attr.value[k].expression.codeChunks[0].includes(`' ${scopedClass}'`)) {
          attr.value[k].expression.codeChunks[0] = `(${attr.value[k].expression.codeChunks[0]}) + ' ${scopedClass}'`;
        }
      }
    }
  }
}
function transformStyles({compileOptions, filename, fileID}) {
  const styleNodes = [];
  const styleTransformPromises = [];
  const scopedClass = `astro-${hashFromFilename(fileID)}`;
  const nodesToScope = new Set();
  return {
    visitors: {
      html: {
        InlineComponent: {
          enter(node) {
            if (node.name === "Markdown") {
              injectScopedClassAttribute(node, scopedClass, "$scope");
            }
            for (let attr of node.attributes) {
              if (attr.name === "class") {
                injectScopedClassAttribute(node, scopedClass, "class");
                break;
              }
            }
          }
        },
        Element: {
          enter(node) {
            if (node.name === "style") {
              const code = Array.isArray(node.children) ? node.children.map(({data}) => data).join("\n") : "";
              if (!code)
                return;
              const langAttr = (node.attributes || []).find(({name}) => name === "lang");
              const globalAttr = (node.attributes || []).find(({name}) => name === "global");
              styleNodes.push(node);
              styleTransformPromises.push(transformStyle(code, {
                logging: compileOptions.logging,
                type: langAttr && langAttr.value[0] && langAttr.value[0].data || void 0,
                filename,
                scopedClass,
                tailwindConfig: compileOptions.astroConfig.devOptions.tailwindConfig,
                global: globalAttr && globalAttr.value
              }));
              return;
            }
            if (NEVER_SCOPED_TAGS.has(node.name) || node.name.toLowerCase() === "!doctype") {
              return;
            }
            nodesToScope.add(node);
          }
        }
      },
      css: {
        Style: {
          enter(node) {
            if (!node.content || !node.content.styles)
              return;
            const code = node.content.styles;
            const langAttr = (node.attributes || []).find(({name}) => name === "lang");
            const globalAttr = (node.attributes || []).find(({name}) => name === "global");
            styleNodes.push(node);
            styleTransformPromises.push(transformStyle(code, {
              logging: compileOptions.logging,
              type: langAttr && langAttr.value[0] && langAttr.value[0].data || void 0,
              filename,
              scopedClass,
              global: globalAttr && globalAttr.value
            }));
          }
        }
      }
    },
    async finalize() {
      const styleTransforms = await Promise.all(styleTransformPromises);
      if (styleTransforms.length > 0) {
        for (const node of nodesToScope.values()) {
          injectScopedClassAttribute(node, scopedClass);
        }
      }
      styleTransforms.forEach((result, n) => {
        if (styleNodes[n].attributes) {
          const isHeadStyle = !styleNodes[n].content;
          if (isHeadStyle) {
            styleNodes[n].children = [__spreadProps(__spreadValues({}, styleNodes[n].children[0]), {data: result.css})];
          } else {
            styleNodes[n].content.styles = result.css;
          }
          const styleTypeIndex = styleNodes[n].attributes.findIndex(({name}) => name === "type");
          if (styleTypeIndex !== -1) {
            styleNodes[n].attributes[styleTypeIndex].value[0].raw = "text/css";
            styleNodes[n].attributes[styleTypeIndex].value[0].data = "text/css";
          } else {
            styleNodes[n].attributes.push({name: "type", type: "Attribute", value: [{type: "Text", raw: "text/css", data: "text/css"}]});
          }
          const styleLangIndex = styleNodes[n].attributes.findIndex(({name}) => name === "lang");
          if (styleLangIndex !== -1)
            styleNodes[n].attributes.splice(styleLangIndex, 1);
        }
      });
    }
  };
}
export {
  transformStyles as default
};
