// astro/dist/build/bundle/js.js
import {fileURLToPath} from "url";
import {rollup} from "rollup";
import {terser} from "rollup-plugin-terser";
import {createBundleStats, addBundleStats} from "../stats.js";
import {IS_ASTRO_FILE_URL} from "../util.js";
import cheerio from "cheerio";
import path from "path";
function collectJSImports(buildState) {
  const imports = new Set();
  for (const id of Object.keys(buildState)) {
    if (buildState[id].contentType === "application/javascript")
      imports.add(id);
  }
  return imports;
}
function pageUrlToVirtualJSEntry(pageUrl) {
  return "astro-virtual:" + pageUrl.replace(/.html$/, "").replace(/^\./, "") + ".js";
}
async function bundleHoistedJS({
  buildState,
  astroConfig,
  logging,
  depTree,
  dist,
  runtime
}) {
  const sortedPages = Object.keys(depTree);
  sortedPages.sort((a, b) => a.localeCompare(b, "en", {numeric: true}));
  const entryImports = [];
  const virtualScripts = new Map();
  const pageToEntryMap = new Map();
  for (let pageUrl of sortedPages) {
    const hoistedJS = depTree[pageUrl].hoistedJS;
    if (hoistedJS.size) {
      for (let [url, scriptInfo] of hoistedJS) {
        if (virtualScripts.has(url) || !url.startsWith("astro-virtual:"))
          continue;
        virtualScripts.set(url, scriptInfo);
      }
      const entryURL = pageUrlToVirtualJSEntry(pageUrl);
      const entryJS = Array.from(hoistedJS.keys()).map((url) => `import '${url}';`).join("\n");
      virtualScripts.set(entryURL, {
        content: entryJS
      });
      entryImports.push(entryURL);
      pageToEntryMap.set(pageUrl, entryURL);
    }
  }
  if (!entryImports.length) {
    return;
  }
  const inputOptions = {
    input: entryImports,
    plugins: [
      {
        name: "astro:build",
        resolveId(source, imported) {
          if (virtualScripts.has(source)) {
            return source;
          }
          if (source.startsWith("/")) {
            return source;
          }
          if (imported) {
            const outUrl = new URL(source, "http://example.com" + imported);
            return outUrl.pathname;
          }
          return null;
        },
        async load(id) {
          if (virtualScripts.has(id)) {
            let info = virtualScripts.get(id);
            return info.content;
          }
          const result = await runtime.load(id);
          if (result.statusCode !== 200) {
            return null;
          }
          return result.contents.toString("utf-8");
        }
      }
    ]
  };
  const build = await rollup(inputOptions);
  const outputOptions = {
    dir: fileURLToPath(dist),
    format: "esm",
    exports: "named",
    entryFileNames(chunk) {
      const {facadeModuleId} = chunk;
      if (!facadeModuleId)
        throw new Error(`facadeModuleId missing: ${chunk.name}`);
      return facadeModuleId.substr("astro-virtual:/".length, facadeModuleId.length - "astro-virtual:/".length - 3) + "-[hash].js";
    },
    plugins: [
      terser()
    ]
  };
  const {output} = await build.write(outputOptions);
  const entryToChunkFileName = new Map();
  output.forEach((chunk) => {
    const {fileName, facadeModuleId, isEntry} = chunk;
    if (!facadeModuleId || !isEntry)
      return;
    entryToChunkFileName.set(facadeModuleId, fileName);
  });
  Object.keys(buildState).forEach((id) => {
    if (buildState[id].contentType !== "text/html")
      return;
    const entryVirtualURL = pageUrlToVirtualJSEntry(id);
    let hasHoisted = false;
    const $ = cheerio.load(buildState[id].contents);
    $('script[data-astro="hoist"]').each((i, el) => {
      hasHoisted = true;
      if (i === 0) {
        let chunkName = entryToChunkFileName.get(entryVirtualURL);
        if (!chunkName)
          return;
        let chunkPathname = "/" + chunkName;
        let relLink = path.relative(path.dirname(id), chunkPathname);
        $(el).attr("src", relLink.startsWith(".") ? relLink : "./" + relLink);
        $(el).removeAttr("data-astro");
        $(el).html("");
      } else {
        $(el).remove();
      }
    });
    if (hasHoisted) {
      buildState[id].contents = $.html();
    }
  });
}
async function bundleJS(imports, {astroRuntime, dist}) {
  const ROOT = "astro:root";
  const validImports = [...imports].filter((url) => IS_ASTRO_FILE_URL.test(url));
  const root = `
  ${validImports.map((url) => `import '${url}';`).join("\n")}
`;
  const inputOptions = {
    input: validImports,
    plugins: [
      {
        name: "astro:build",
        resolveId(source, imported) {
          if (source === ROOT) {
            return source;
          }
          if (source.startsWith("/")) {
            return source;
          }
          if (imported) {
            const outUrl = new URL(source, "http://example.com" + imported);
            return outUrl.pathname;
          }
          return null;
        },
        async load(id) {
          if (id === ROOT) {
            return root;
          }
          const result = await astroRuntime.load(id);
          if (result.statusCode !== 200) {
            return null;
          }
          return result.contents.toString("utf-8");
        }
      }
    ]
  };
  const build = await rollup(inputOptions);
  const outputOptions = {
    dir: fileURLToPath(dist),
    format: "esm",
    exports: "named",
    entryFileNames(chunk) {
      const {facadeModuleId} = chunk;
      if (!facadeModuleId)
        throw new Error(`facadeModuleId missing: ${chunk.name}`);
      return facadeModuleId.substr(1);
    },
    plugins: [
      terser()
    ]
  };
  const stats = createBundleStats();
  const {output} = await build.write(outputOptions);
  await Promise.all(output.map(async (chunk) => {
    const code = chunk.code || "";
    await addBundleStats(stats, code, chunk.fileName);
  }));
  return stats;
}
export {
  bundleHoistedJS,
  bundleJS,
  collectJSImports
};
