// astro/dist/compiler/transform/postcss-scoped-styles/index.js
const GLOBAL = ":global(";
const CSS_SEPARATORS = new Set([" ", ",", "+", ">", "~"]);
const KEYFRAME_PERCENT = /\d+\.?\d*%/;
function minifySelector(selector) {
  return selector.replace(/(\r?\n|\s)+/g, " ").replace(/\s*(,|\+|>|~|\(|\))\s*/g, "$1");
}
function matchParen(search, start) {
  if (search[start] !== "(")
    return -1;
  let parenCount = 0;
  for (let n = start + 1; n < search.length; n++) {
    if (search[n] === ")" && parenCount === 0)
      return n;
    if (search[n] === "(")
      parenCount += 1;
    if (search[n] === ")")
      parenCount -= 1;
  }
  return -1;
}
const NEVER_SCOPED_TAGS = new Set(["base", "body", "font", "frame", "frameset", "head", "html", "link", "meta", "noframes", "noscript", "script", "style", "title"]);
function scopeRule(selector, className) {
  if (selector === "from" || selector === "to" || KEYFRAME_PERCENT.test(selector)) {
    return selector;
  }
  const input = minifySelector(selector);
  const c = className.replace(/^\.?/, ".");
  const selectors = [];
  let ss = input;
  {
    let start = 0;
    let lastValue = "";
    let parenCount = 0;
    for (let n = 0; n < ss.length; n++) {
      const isEnd = n === input.length - 1;
      if (input[n] === "(")
        parenCount += 1;
      if (input[n] === ")")
        parenCount -= 1;
      if (isEnd || parenCount === 0 && CSS_SEPARATORS.has(input[n])) {
        lastValue = input.substring(start, isEnd ? void 0 : n);
        if (!lastValue)
          continue;
        selectors.push({start, end: isEnd ? n + 1 : n, value: lastValue});
        start = n + 1;
      }
    }
  }
  for (let i = selectors.length - 1; i >= 0; i--) {
    const {start, end} = selectors[i];
    let value = selectors[i].value;
    const head = ss.substring(0, start);
    const tail = ss.substring(end);
    if (value.includes(GLOBAL)) {
      let withoutGlobal = value;
      while (withoutGlobal.includes(GLOBAL)) {
        const globalStart = withoutGlobal.indexOf(GLOBAL);
        const globalParenOpen = globalStart + GLOBAL.length - 1;
        const globalEnd = matchParen(withoutGlobal, globalParenOpen);
        const globalContents = withoutGlobal.substring(globalParenOpen + 1, globalEnd);
        withoutGlobal = withoutGlobal.substring(0, globalStart) + globalContents + withoutGlobal.substring(globalEnd + 1);
      }
      ss = head + withoutGlobal + tail;
      continue;
    }
    if (value.includes("*")) {
      ss = head + value.replace(/\*/g, c) + tail;
      continue;
    }
    if (CSS_SEPARATORS.has(value) || NEVER_SCOPED_TAGS.has(value)) {
      ss = head + value + tail;
      continue;
    }
    let pseudoclassStart = -1;
    for (let n = 0; n < value.length; n++) {
      if (value[n] === ":" && value[n - 1] !== "\\") {
        pseudoclassStart = n;
        break;
      }
    }
    if (pseudoclassStart !== -1) {
      ss = head + value.substring(0, pseudoclassStart) + c + value.substring(pseudoclassStart) + tail;
    } else {
      ss = head + value + c + tail;
    }
  }
  return ss;
}
function astroScopedStyles(options) {
  const rulesScopedCache = new WeakSet();
  return {
    postcssPlugin: "@astrojs/postcss-scoped-styles",
    Rule(rule) {
      if (!rulesScopedCache.has(rule)) {
        rule.selector = scopeRule(rule.selector, options.className);
        rulesScopedCache.add(rule);
      }
    }
  };
}
export {
  NEVER_SCOPED_TAGS,
  astroScopedStyles as default,
  scopeRule
};
