// astro/dist/compiler/transform/util/end-of-head.js
const beforeHeadElements = new Set(["!doctype", "html"]);
const validHeadElements = new Set(["title", "meta", "link", "style", "script", "noscript", "base"]);
class EndOfHead {
  constructor() {
    this.html = null;
    this.head = null;
    this.firstNonHead = null;
    this.parent = null;
    this.stack = [];
    this.foundHeadElements = false;
    this.foundBodyElements = false;
    this.append = () => void 0;
  }
  get found() {
    return !!(this.head || this.firstNonHead);
  }
  get foundHeadContent() {
    return !!this.head || this.foundHeadElements;
  }
  get foundHeadAndBodyContent() {
    return this.foundHeadContent && this.foundBodyElements;
  }
  get foundHeadOrHtmlElement() {
    return !!(this.html || this.head);
  }
  enter(node) {
    const name = node.name ? node.name.toLowerCase() : null;
    if (this.found) {
      if (!validHeadElements.has(name)) {
        if (node.type === "Element") {
          this.foundBodyElements = true;
        }
      }
      return;
    }
    this.stack.push(node);
    if (!node.name) {
      return;
    }
    if (name === "head") {
      this.head = node;
      this.parent = this.stack[this.stack.length - 2];
      this.append = this.appendToHead;
      return;
    }
    if (beforeHeadElements.has(name)) {
      if (name === "html") {
        this.html = node;
      }
      return;
    }
    if (!validHeadElements.has(name)) {
      if (node.type === "Element") {
        this.foundBodyElements = true;
      }
      this.firstNonHead = node;
      this.parent = this.stack[this.stack.length - 2];
      this.append = this.prependToFirstNonHead;
      return;
    } else {
      this.foundHeadElements = true;
    }
  }
  leave(_node) {
    this.stack.pop();
  }
  appendToHead(...nodes) {
    var _a;
    if (this.head) {
      const head = this.head;
      head.children = (_a = head.children) != null ? _a : [];
      head.children.push(...nodes);
    }
  }
  prependToFirstNonHead(...nodes) {
    var _a, _b, _c, _d;
    let idx = this.firstNonHead && ((_b = (_a = this.parent) == null ? void 0 : _a.children) == null ? void 0 : _b.indexOf(this.firstNonHead)) || 0;
    (_d = (_c = this.parent) == null ? void 0 : _c.children) == null ? void 0 : _d.splice(idx, 0, ...nodes);
  }
}
export {
  EndOfHead
};
