// astro/dist/compiler/utils.js
function isCustomElementTag(tag) {
  return /[-]/.test(tag);
}
function isComponentTag(tag) {
  return /^[A-Z]/.test(tag) || /^[a-z]+\./.test(tag) || isCustomElementTag(tag);
}
function clamp(num, min, max) {
  return Math.max(min, Math.min(max, num));
}
function positionAt(offset, text) {
  offset = clamp(offset, 0, text.length);
  const lineOffsets = getLineOffsets(text);
  let low = 0;
  let high = lineOffsets.length;
  if (high === 0) {
    return {line: 0, character: offset};
  }
  while (low < high) {
    const mid = Math.floor((low + high) / 2);
    if (lineOffsets[mid] > offset) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }
  const line = low - 1;
  return {line, character: offset - lineOffsets[line]};
}
function offsetAt(position, text) {
  const lineOffsets = getLineOffsets(text);
  if (position.line >= lineOffsets.length) {
    return text.length;
  } else if (position.line < 0) {
    return 0;
  }
  const lineOffset = lineOffsets[position.line];
  const nextLineOffset = position.line + 1 < lineOffsets.length ? lineOffsets[position.line + 1] : text.length;
  return clamp(nextLineOffset, lineOffset, lineOffset + position.character);
}
function getLineOffsets(text) {
  const lineOffsets = [];
  let isLineStart = true;
  for (let i = 0; i < text.length; i++) {
    if (isLineStart) {
      lineOffsets.push(i);
      isLineStart = false;
    }
    const ch = text.charAt(i);
    isLineStart = ch === "\r" || ch === "\n";
    if (ch === "\r" && i + 1 < text.length && text.charAt(i + 1) === "\n") {
      i++;
    }
  }
  if (isLineStart && text.length > 0) {
    lineOffsets.push(text.length);
  }
  return lineOffsets;
}
export {
  clamp,
  isComponentTag,
  isCustomElementTag,
  offsetAt,
  positionAt
};
