// astro/dist/internal/__astro_component.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {enumerable: true, configurable: true, writable: true, value}) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
import hash from "shorthash";
import {valueToEstree} from "estree-util-value-to-estree";
import {generate, GENERATOR} from "astring";
import * as astroHtml from "./renderer-html";
const customGenerator = __spreadProps(__spreadValues({}, GENERATOR), {
  Literal(node, state) {
    if (node.raw != null) {
      state.write(node.raw.replace("</script>", "<\\/script>"));
    } else {
      GENERATOR.Literal(node, state);
    }
  }
});
const serialize = (value) => generate(valueToEstree(value), {
  generator: customGenerator
});
const astroHtmlRendererInstance = {
  name: null,
  source: "",
  renderer: astroHtml,
  polyfills: [],
  hydrationPolyfills: []
};
let rendererInstances = [];
function setRenderers(_rendererInstances) {
  rendererInstances = [].concat(_rendererInstances);
}
function isCustomElementTag(name) {
  return typeof name === "string" && /-/.test(name);
}
const rendererCache = new Map();
function inferClientRenderer(metadata) {
  if (rendererInstances.length === 1) {
    return rendererInstances[0];
  } else if (metadata.value) {
    const hint = metadata.value;
    let match = rendererInstances.find((instance) => instance.name === hint);
    if (!match) {
      const fullHintName = `@astrojs/renderer-${hint}`;
      match = rendererInstances.find((instance) => instance.name === fullHintName);
    }
    if (!match) {
      throw new Error(`Couldn't find a renderer for <${metadata.displayName} client:only="${metadata.value}" />. Is there a renderer that matches the "${metadata.value}" hint in your Astro config?`);
    }
    return match;
  } else {
    throw new Error(`Can't determine the renderer for ${metadata.displayName}. Include a hint similar to <${metadata.displayName} client:only="react" /> when multiple renderers are included in your Astro config.`);
  }
}
async function resolveRenderer(Component, props = {}, children, metadata = {}) {
  if (metadata.hydrate === "only") {
    return inferClientRenderer(metadata);
  }
  if (rendererCache.has(Component)) {
    return rendererCache.get(Component);
  }
  const errors = [];
  for (const instance of rendererInstances) {
    const {renderer} = instance;
    try {
      const shouldUse = await renderer.check(Component, props, children);
      if (shouldUse) {
        rendererCache.set(Component, instance);
        return instance;
      }
    } catch (err) {
      errors.push(err);
    }
  }
  if (errors.length) {
    throw errors[0];
  }
}
async function generateHydrateScript(scriptOptions, metadata) {
  const {instance, astroId, props} = scriptOptions;
  const {source} = instance;
  const {hydrate, componentUrl, componentExport} = metadata;
  let hydrationSource = "";
  if (instance.hydrationPolyfills.length) {
    hydrationSource += `await Promise.all([${instance.hydrationPolyfills.map((src) => `import("${src}")`).join(", ")}]);
`;
  }
  hydrationSource += source ? `
  const [{ ${componentExport.value}: Component }, { default: hydrate }] = await Promise.all([import("${componentUrl}"), import("${source}")]);
  return (el, children) => hydrate(el)(Component, ${serialize(props)}, children);
` : `
  await import("${componentUrl}");
  return () => {};
`;
  const hydrationScript = `<script type="module">
import setup from '/_astro_frontend/hydrate/${hydrate}.js';
setup("${astroId}", {${metadata.value ? `value: "${metadata.value}"` : ""}}, async () => {
  ${hydrationSource}
});
</script>`;
  return hydrationScript;
}
const getComponentName = (Component, componentProps) => {
  var _a;
  if (componentProps.displayName)
    return componentProps.displayName;
  switch (typeof Component) {
    case "function":
      return (_a = Component.displayName) != null ? _a : Component.name;
    case "string":
      return Component;
    default: {
      return Component;
    }
  }
};
const prepareSlottedChildren = (children) => {
  const $slots = {
    default: ""
  };
  for (const child of children) {
    if (typeof child === "string") {
      $slots.default += child;
    } else if (typeof child === "object" && child["$slot"]) {
      if (!$slots[child["$slot"]])
        $slots[child["$slot"]] = "";
      $slots[child["$slot"]] += child.children.join("").replace(new RegExp(`slot="${child["$slot"]}"s*`, ""));
    }
  }
  return {$slots};
};
const removeSlottedChildren = (_children) => {
  let children = "";
  for (const child of _children) {
    if (typeof child === "string") {
      children += child;
    } else if (typeof child === "object" && child["$slot"]) {
      children += child.children.join("");
    }
  }
  return children;
};
function __astro_component(Component, metadata = {}) {
  if (Component == null) {
    throw new Error(`Unable to render ${metadata.displayName} because it is ${Component}!
Did you forget to import the component or is it possible there is a typo?`);
  } else if (typeof Component === "string" && !isCustomElementTag(Component)) {
    throw new Error(`Astro is unable to render ${metadata.displayName}!
Is there a renderer to handle this type of component defined in your Astro config?`);
  }
  return async function __astro_component_internal(props, ..._children) {
    if (Component.isAstroComponent) {
      return Component.__render(props, prepareSlottedChildren(_children));
    }
    const children = removeSlottedChildren(_children);
    let instance = await resolveRenderer(Component, props, children, metadata);
    if (!instance) {
      if (isCustomElementTag(Component)) {
        instance = astroHtmlRendererInstance;
      } else {
        instance = rendererInstances.length === 2 ? rendererInstances[1] : void 0;
      }
      if (!instance) {
        const name = getComponentName(Component, metadata);
        throw new Error(`No renderer found for ${name}! Did you forget to add a renderer to your Astro config?`);
      }
    }
    let html = "";
    if (metadata.hydrate !== "only") {
      const rendered = await instance.renderer.renderToStaticMarkup(Component, props, children, metadata);
      html = rendered.html;
    }
    if (instance.polyfills.length) {
      let polyfillScripts = instance.polyfills.map((src) => `<script type="module" src="${src}"></script>`).join("");
      html = html + polyfillScripts;
    }
    if (!metadata.hydrate) {
      return html.replace(/\<\/?astro-fragment\>/g, "");
    }
    const uniqueId = props[Symbol.for("astro.context")].createAstroRootUID(html);
    const uniqueIdHashed = hash.unique(uniqueId);
    const script = await generateHydrateScript({instance, astroId: uniqueIdHashed, props}, metadata);
    const astroRoot = `<astro-root uid="${uniqueIdHashed}">${html}</astro-root>`;
    return [astroRoot, script].join("\n");
  };
}
export {
  __astro_component,
  setRenderers
};
